README.md
<img src=https://raw.githubusercontent.com/Art-Institute-of-Chicago/template/master/aic-logo.gif alt="Art Institute of Chicago" width=20% />
<img src=journeymaker-logo.png alt="JourneyMaker: Your Journey Begins Here" width=75% style="float: right"/>

# JourneyMaker Client
> A JavaScript app that digests data from the JourneyMaker CMS

![Screenshot of JourneyMaker](/screenshot.png?raw=true)

[JourneyMaker](http://www.artic.edu/journeymaker/) is an innovative new interactive experience that empowers families to create
their own personalized journeys of the Art Institute of Chicago. Designed as a web application
and developed for multi-touch screens, the interactive is allows your family to create your
very own tour of the museum. Choose one of eight storylines—like Superheroes, Time Travelers, or Strange
and Wild Creatures—and then select works from the museum’s collection within that story.
After you’ve made your selections, print out your personalized Journey Guide, which includes
information, activities, and wayfinding directions. Then head into the museum for an art-filled
adventure!

JourneyMaker launched at the Art Institute of Chicago on July 1, 2016 as six multi-touchscreens in our Ryan 
Learning Center, along with a [desktop version](http://journeymaker.artic.edu/) available online. It's 
still in place today and is being maintained by a team of internal developers.



## Features

* Administered via a Drupal CMS
* Interactive story and object selector
* Powered by TUIO for accurate multi-touch
* Dynamically accounts for objects that are off-view
* Wayfinding algorithm optimizes tour path, accounting for elevation
* Print personalized tours via in-gallery printers or generate PDFs



## Overview

JourneyMaker consists of two parts: a content authoring system written in Drupal and a public-facing frontend written in JavaScript. This repository contains the client code. In order to deploy your own instance of JourneyMaker, you will also need to install the JourneyMaker CMS:

https://github.com/art-institute-of-chicago/journeymaker-cms

The JourneyMaker CMS does not need to be installed on the same server or under the same domain as the client: as long as the client can access the assets stored in the CMS over HTTP ([CORS](http://stackoverflow.com/questions/20035101/no-access-control-allow-origin-header-is-present-on-the-requested-resource)), everything should work fine. For cross-domain development, you can use the [Allow-Control-Allow-Origin: \*](https://chrome.google.com/webstore/detail/allow-control-allow-origi/nlfbmbojpeacfghkpbjhddihlkkiljbi) Chrome extension.

The JourneyMaker client can run in two modes: kiosk or web app. The web app is meant to be accessed via HTTP as any website. Kiosk mode is meant to enable its deployment as an in-gallery interactive.

For brevity, throughout the rest of this repo's documentation, "JourneyMaker" refers to the public-facing frontend client, i.e. this repository. "JourneyMaker CMS" refers to the Drupal instance you've installed from the link above.



## Requirements

JourneyMaker can run in two modes: kiosk or webapp. For webapp mode, any webserver (Apache, nginx, etc.) is sufficient. Kiosk mode, however, requires Node.js v0.12.7: [kiosk mode requirements](README.KIOSK.md#requirements).

Due to its reliance on shell scripts, JourneyMaker requires a \*NIX environment (Linux, Mac OS X, etc.). On Windows, [Git Bash](https://git-for-windows.github.io/) should work fine, but is not officially supported.

We use Mac OS X to develop and deploy, so our code and documentation is biased towards that OS. Feel free to report any issues you encounter while running JourneyMaker on other \*NIX environments.



### Fonts and Sounds

We use licensed materials in this interactive which are proprietary to the Art Institute of Chicago and/or are licensed from third-party distributors. These materials are not included as part of this repository.

This project utilizes the third-party font Anchor Medium. You may [purchase your own license](https://processtypefoundry.com/fonts/anchor/complete-specs-webfonts) or replace references to this font with one of your own.

This project utilizes the font Ideal Sans. This is an Art Institute proprietary font owned by the Art Institute of Chicago. You should replace references to this font with one of your own.

Fonts are stored in `App/resources/fonts`.

This project utilizes sounds licensed from [AudioJungle](https://audiojungle.net). The audio clips located in `App/resources/sfx` have been replaced with dummy silent audio clips. You can replace these audio files with sound effects of your choosing, or purchase the sounds we used:

https://audiojungle.net/item/menu-button-sounds-pack/4023594?s_rank=1
https://audiojungle.net/item/wood-click/11845976?s_rank=2
https://audiojungle.net/item/elegant-button-clicks/12286426?s_rank=1



## Installation

We created an `install.sh` script that will walk you through the setup process.

```bash
# Clone the repo to your computer
git clone https://github.com/art-institute-of-chicago/journeymaker-client

# Enter the folder that was created by the clone
cd journeymaker-client

# Run the install script with sudo privilege
sudo bash install.sh

# If you do not have sudo rights, try this instead:
chmod a+rx install.sh
./install.sh

```

The script will ask you if you'd like to install JourneyMaker in webapp or kiosk mode. It will then create an `App/config.auto.json` with the necessary settings toggled to support that mode. On its first run, it will also create an `App/config.custom.json`, which you must edit manually.

It is safe to re-run this script multiple times. It will not erase any of your custom settings.

If you encounter issues with the install script, check the `install.log` file in the root of your cloned repo.



### Webapp Mode

In webapp mode, JourneyMaker can run on any web server and does not require node to function. We typically serve it using Apache. The recommended deployment process is as follows:

1. Install JourneyMaker as described above
2. Set your webserver to point to the `App` subdirectory
3. Run the install script and type `1` for `1) Webapp`
4. Edit `config.custom.json` according to the Configuration section.

In particular, you must ensure that `contentOrigin` in your custom config is set to the URL of `json.json` generated by your JourneyMaker CMS instance.

The webapp requires an HTML-to-PDF generator to work. We are working on open-sourcing the PHP code for our remote generator, but for now, you may use our generator endpoint for your JourneyMaker projects. See `webPdfGeneratorPath` for more details.



## Configuration

Check the `App` folder after running `install.sh`. There should be three config files:

```
App/config.auto.json = generated whenever you run install.sh
App/config.custom.json = edit this file!
App/config.defaults.json = lists available configs and their defaults
```

These configs will be applied in a sort of "stack" – configs in `auto` will override `custom`, which will in turn override `defaults`. Whenever you need to change a config setting, such as `contentOrigin` or `analyticsId`, please do so in `config.custom.json`.

Note that `config.auto.json` will be overriden everytime you run `install.sh` – its purpose is merely to toggle between kiosk and web modes. Likewise, please do not edit `config.defaults.json`, since it might be overriden by future updates to this repository.

Here is a sample `config.custom.json` explaining the various settings:

```javascript
{

    // IMPORTANT: Change this to point at your webserver, specifically
    //   to wherever the App folder's contents can be accessed via HTTP.
    // Please do not include a trailing slash!
    // This defaults to Apache's default settings:
    "baseURL": "http://localhost:80",

    // If you must host JourneyMaker in a subdirectory, try something like:
    // "baseURL": "https://your.museum/journeymaker",

    // For kiosk mode: AssetDownloader prepends the baseURL to the URLs of
    //   assets it downloads. If you change the baseURL, you will have to
    //   re-run the AssetDownloader.

    // IMPORTANT: Change this to point at your JourneyMaker CMS.
    // The webapp mode queries contentOrigin directly on each visit.
    // The kiosk mode's AssetDownloader caches this file.
    "contentOrigin": "http://journeymaker.your.museum/sites/default/files/json/json.json",

    // The interactive will reset if left alone for this many seconds
    "timeoutSecs": 180,

    // This is relevant to kiosk only. JourneyMaker defaults to accept
    //   mouse input; TUIO is a multi-touch protocol. If you'd like to
    //   use a touchscreen in your kiosk, we recommend disabling mouse
    //   and enabling TUIO. You will need to install and run a UPDD
    //   TUIO Server, as well as the Node.js TUIOServer included in this
    //   repo to make it work. See README.KIOSK.md for more details.
    // This example shows the defaults. Please don't enable both mouse
    //   and TUIO input at the same time, effects of this are unknown.
    "enableMouse": true,
    "enableTuio": false,

    // This only applies in kiosk mode.
    // This is where the PrintServer will save the PDFs it generates.
    // PrintServer for kiosk != remote PDF generator for webapp!
    // You can set this to whatever you'd like, default is "/tmp"
    // Note that the path is relative to the PrintServer folder;
    //   ~/ doesn't seem to work, type out /Users/user/ etc.
    "printOutputPath": "../Kiosk/JourneyGuidePDFs/",

    // This only applies in kiosk mode.
    // `name` must match a printer installed on the system.
    // Try running `lpstat -p` to see available printers.
    // You can add as many printers as you'd like.
    // See README.KIOSK.md for more details.
    "printValidPrinters": [
        {
            "name": "Print_Station_A",
            "displayName": "Print Station A",
            "color": "#009bce"
        }
    ],

    // This only applies in kiosk mode. Defaults to "random".
    // Set this to the `name` of a printer listed in `printValidPrinters`.
    // If it does not match the `name` of any printer, printer will be
    //   picked randomly from the list of valid printers.
    // See README.KIOSK.md for more details.
    "printPreferredPrinter": "random",

    // We are working on opensourcing our remote PDF generator code.
    // For the time being, you may use our remote PDF generator.
    // In fact, install.sh sets it as the default for this setting:
    "webPdfGeneratorPath": "http://journeymaker.artic.edu/PDF/phppdf.php",

    // During the "Creating Your Path..." sorting stage,
    //   objects that are not located in any of these galleries
    //   will be pushed to appear later in the sorting order.
    // We use this to ensure that objects nearest the in-gallery
    //   printers (i.e. in the Modern Wing) will be sorted first.
    "priorityGalleryIds": [
        23965, 23966, 23967, 23969, 23970, 23971,
        23972, 23973, 23974, 23975, 23976, 23977,
        23978, 23989, 23990, 23991, 23992, 23993,
        23994, 23995, 23996, 23997, 23999, 24000,
        24300, 24301, 24302, 24303, 24304, 24305,
        24306, 24307, 24308, 24309, 24310, 24311,
        24312, 24313, 24317, 24318, 24366, 24563,
        24564, 24565, 24566, 24568, 24569, 24570,
        24571, 24572, 24573, 24575, 25208, 25210,
        25456, 25467, 25468, 25469, 25470, 25471,
        25472, 25473, 25474, 25475, 25476, 25477,
        25478, 25479, 25490, 26509, 26771
    ],

    // This is used to sort the themes provided by the CMS
    // Themes that don't appear in this list will still be displayed
    // This only determines sort order
    "themeIds": [ 444, 403, 612, 361, 484, 524, 565, 703 ],

    // This is also used during the "Creating Your Path..." sort.
    // There is an algorithm that determines the optimal order,
    //  based on your starting point and distance between objects,
    //  accounting for what floor they are on and the height of each floor.
    // Set this to the latitute and longitude of your kiosk.
    "guideStartX": 41.880591,
    "guideStartY": -87.62172,
    "guideStartFloor": 1,
    "guideFloorOffsetMeters": 135,

    // Google Analytics settings, replace as needed
    "enableAnalytics": true,
    "analyticsId": "UA-XXXXXXX-XX",
    // This is the prefix we use to build our GA views
    // You can modify install.sh to set different prefixes for each mode
    "analyticsPrefix": "journey-maker/home-companion/",
    "analyticsDebug": false,

    // Replace this with you museum's admissions and visit pages
    "buyTicketsUrl": "https://sales.artic.edu/",
    "visitUrl": "http://www.artic.edu/visit/visiting-your-family"

}
```

You may override any setting from `config.defaults.json` in your `config.custom.json`. Please search the codebase to see what these settings do. Additionally, please see [README.KIOSK.md](README.KIOSK.md) for information regarding the kiosk mode settings.



## Contributing

We encourage your contributions. Please fork this repository and make your changes in a separate branch. 
We like to use [git-flow](https://github.com/nvie/gitflow) to make this process easier.

```bash
# Clone the repo to your computer
git clone https://github.com/your-github-account/journeymaker-client.git

# Enter the folder that was created by the clone
cd journeymaker-client

# Run the install script
chmod a+rx install.sh
./install.sh

# Start a feature branch
git flow start feature yourinitials-good-description-issuenumberifapplicable

# ... make some changes, commit your code

# Push your branch to GitHub
git push origin yourinitials-good-description-issuenumberifapplicable
```

Then on github.com, create a Pull Request to merge your changes into our 
`develop` branch. 

This project is released with a Contributor Code of Conduct. By participating in 
this project you agree to abide by its [terms](CODE_OF_CONDUCT.md).

We also welcome bug reports and questions under GitHub's [Issues](issues).



## Acknowledgements

Design and Development by [Belle & Wissell, Co.](http://www.bwco.info/)



## Licensing

The code in this project is licensed under the AGNU Affero General Public License Version 3.


CHANGELOG.md
JourneyMaker Client Changelog
=============================

At the Digital Experience department of the Art Institute of Chicago, we use Redmine as our internal issue tracker. Some of the changes listed here reference issues tracked via Redmine. In order to differentiate between references to GitHub issues and Redmine issues, we are adopting the following scheme:

```
* Change description (#14) [#23]

(#14) = reference to a GitHub issue
[#23] = reference to a Redmine issue
```

Outside contributors can ignore Redmine references.

-----------------------------

1.2.0 - Multiple language support (in progress by Belle & Wissel Co.)

1.1.1 - Tweaks to documentation

* Update spacing of images in README
* Make corrections to README

1.1.0 - Changes for open source release:

* Lint whitespace for consistency [#1462]
* Improve documentation on deployment process [#1480]
* Remove AIC-specific deployment-related items [#1477]
* Remove old SVN repository data from LogServer [#1479]
* Move the contents of the `journeymaker` subfolder into root [#1475]
* Modify `ConfigModel.load()` to accept multiple files [#1464]
* Create `config.custom.json`, which extends and overrides `config.default.json` on a per-setting basis [#1463]
* Create `package.json` for LogServer, PrintServer, and TUIOServer [#1469, #1457]
* Write `install.sh` to run `npm install`, generate config, and other tasks [#1468, #1478]
* Add downloaded assets and `node_modules` to `.gitignore` [#1461, #1470]
* Add README.md and README.KIOSK.md [#1485, #1486, #1522]
* Add CHANGELOG.md [#1506]

1.0.0 - In-gallery deployment (initial)


install.sh
#!/bin/bash

# http://stackoverflow.com/questions/29436275
function yes_or_no {
	while true; do
		read -p "$* [y/n]: " yn
		case $yn in
			[Yy]*) return 0  ;;
			[Nn]*) echo "Aborted" ; return  1 ;;
		esac
	done
}

# Color markup
R=`tput setaf 1`
G=`tput setaf 2`
Y=`tput setaf 3`
C=`tput setaf 6`
W=`tput setaf 7`

# Get location of this script
DIR_REPO="$( cd "$(dirname $0)" ; pwd -P )"

# Useful if code is stored in a subfolder
DIR_ROOT="$DIR_REPO"

DIR_APP="$DIR_ROOT/App"
DIR_DOWNLOADER="$DIR_ROOT/AssetDownloader"
DIR_LOGGER="$DIR_ROOT/LogServer"
DIR_PRINTER="$DIR_ROOT/PrintServer"
DIR_TUIO="$DIR_ROOT/TUIOServer"

LOG="$DIR_ROOT/install.log"

# Determine if the app is to be installed in webapp or kiosk mode

echo "${C}Welcome to the JourneyMaker Client installer.${W}"
echo "${G}Please select the deployment mode:${W}"

PS3="${W}Mode [1-3]: ${W}"
options=("Webapp" "Kiosk" "Quit")
select opt in "${options[@]}"
do
	case $opt in
		"Webapp")
			MODE="webapp"
			break
		;;
		"Kiosk")
			MODE="kiosk"
			break
		;;
		"Quit")
			echo "${G}Exiting installer...${W}"
			exit
			break
		;;
		*)
			echo "${R}Invalid option. Please enter a number (1-3).${W}"
		;;
	esac
done


echo "${G}Installing in '$MODE' mode...${W}"

# Clear the log
echo "${G}Clearing old install log...${W}"
rm -f "$LOG"
touch "$LOG"
chown $(logname) "$LOG"

if [ ! -f "$DIR_APP/config.custom.json" ]; then

	echo "${W}No existing config.custom.json found, one will be created${W}"
	echo "${G}Performing first-time setup...${W}"

	# Default options
	OPTIONS=`cat <<- EOF
	{
	    "baseURL": null,

	    "contentOrigin": null,

	    "timeoutSecs": 180,

	    "printOutputPath": "/tmp",

	    "printValidPrinters": [
	        {
	            "name": "Print_Station_A",
	            "displayName": "Print Station A",
	            "color": "#009bce"
	        },
	        {
	            "name": "Print_Station_B",
	            "displayName": "Print Station B",
	            "color": "#dd4444"
	        }
	    ],

	    "printPreferredPrinter": "random",

	    "priorityGalleryIds": null,
	    "themeIds": null,

	    "enableAnalytics": false,
	    "analyticsId": null,
	    "analyticsPrefix": null,

	    "guideStartX": 41.880591,
	    "guideStartY": -87.62172,
	    "guideStartFloor": 1,
	    "guideFloorOffsetMeters": 60,

	    "buyTicketsUrl": "javascript:;",
	    "visitUrl": "javascript:;"

	}
	EOF`

	echo "$OPTIONS" > "$DIR_APP/config.custom.json"
	chown $(logname) "$DIR_APP/config.custom.json"

	echo "${W}Created stub config.custom.json${W}"
fi


# Handle webapp installation

if [ "$MODE" == "webapp" ]; then

	# Default options
	OPTIONS=`cat <<- EOF
	{
	    "isHomeCompanion": true,
	    "contentPath": null,
	    "enableMouse": true,
	    "enableLogging": false,
	    "enableTuio": false,
	    "enablePrintSocket": false,
	    "enableWebPdfGeneration": true,
	    "webPdfGeneratorPath": "http://journeymaker.artic.edu/PDF/phppdf.php"
	}
	EOF`


	echo "$OPTIONS" > "$DIR_APP/config.auto.json"
	chown $(logname) "$DIR_APP/config.auto.json"

	echo "${W}Toggled webapp mode in config.auto.json${W}"

fi

# Handle kiosk installation

if [ "$MODE" == "kiosk" ]; then

	# Check requirements
	command -v node >/dev/null 2>&1 || { echo >&2 "${R}node is missing. Aborting.${W}"; exit 1; }
	command -v npm >/dev/null 2>&1 || { echo >&2 "${R}npm is missing. Aborting.${W}"; exit 1; }

	# Check node version
	if [[ ! $(node -v) == *"v0.12"* ]]; then
		echo "${Y}You are using node $(node -v). npm install may fail.${W}"
		echo "${Y}Only node v0.12.x (LTS) is officially supported.${W}"
	fi

	# Default options
	OPTIONS=`cat <<- EOF
	{
	    "isHomeCompanion": false,
	    "contentPath": "assets/data.json",
	    "enableLogging": true,
	    "enablePrintSocket": true,
	    "enableWebPdfGeneration": false
	}
	EOF`

	echo "$OPTIONS" > "$DIR_APP/config.auto.json"
	chown $(logname) "$DIR_APP/config.auto.json"

	echo "${W}Toggled kiosk mode in config.auto.json${W}"

	# Clear old dependencies
	echo "${G}Removing old dependencies...${W}"
	cd "$DIR_DOWNLOADER" && rm -rf "node_modules" >> "$LOG"
	cd "$DIR_LOGGER" && rm -rf "node_modules" >> "$LOG"
	cd "$DIR_PRINTER" && rm -rf "node_modules" >> "$LOG"
	cd "$DIR_TUIO" && rm -rf "node_modules" >> "$LOG"

	# Install npm dependencies
	echo "${G}Installing AssetDownloader dependencies...${W}"
	cd "$DIR_DOWNLOADER" && npm install >> "$LOG"

	echo "${G}Installing LogServer dependencies...${W}"
	cd "$DIR_LOGGER" && npm install >> "$LOG"

	echo "${G}Installing PrintServer dependencies...${W}"
	cd "$DIR_PRINTER" && npm install >> "$LOG"

	echo "${G}Installing TUIOServer dependencies...${W}"
	cd "$DIR_TUIO" && npm install >> "$LOG"

	echo "${G}All node_modules downloaded${W}"
fi

echo "${G}JourneyMaker Client install complete!${W}"

echo "${G}Congrats! You've successfully installed the JourneyMaker Client. You need to adjust some${W}"
echo "${G}configurations personally in App/config.custom.json, in particular the baseURL and${W}"
echo "${G}contentOrigin. See README.md and README.KIOSK.md for the config options.${W}"
if [ "$MODE" == "kiosk" ]; then
    echo
    echo "${G}Since you've installed in kiosk mode, you need to run the following command to store the${W}"
    echo "${G}CMS assets locally:${W}"
    echo
    echo "${G}   chmod a+rx $DIR_DOWNLOADER/download_assets.command${W}"
    echo "${G}   .$DIR_DOWNLOADER/download_assets.command${W}"
fi

echo
echo "${G}Then, point the document root of your web server to the journeymaker-client/App folder and${W}"
echo "${G}your client site should be up and running.${W}"
